components/ndatasource/org.wso2.carbon.ndatasource.ui/src/main/java/org/wso2/carbon/ndatasource/ui/NDataSourceHelper.java

```diff
@@ -595,4 +598,27 @@ public static String getRDBMSEngine(String jdbcUrl) {
         }
         return NDataSourceClientConstants.RDBMSEngines.GENERIC;
     }
+
+	private static DocumentBuilderFactory getSecuredDocumentBuilder() {
+		final int ENTITY_EXPANSION_LIMIT = 0;
+		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+		dbf.setNamespaceAware(true);
+		dbf.setXIncludeAware(false);
+		dbf.setExpandEntityReferences(false);
+
+		try {
+			dbf.setFeature(Constants.SAX_FEATURE_PREFIX + Constants.EXTERNAL_GENERAL_ENTITIES_FEATURE, false);
+			dbf.setFeature(Constants.SAX_FEATURE_PREFIX + Constants.EXTERNAL_PARAMETER_ENTITIES_FEATURE, false);
+			dbf.setFeature(Constants.XERCES_FEATURE_PREFIX + Constants.LOAD_EXTERNAL_DTD_FEATURE, false);
+			dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+		} catch (ParserConfigurationException e) {
+			log.error("Failed to load XML Processor Feature " + Constants.EXTERNAL_GENERAL_ENTITIES_FEATURE + " or "
+					+ Constants.EXTERNAL_PARAMETER_ENTITIES_FEATURE + " or " + Constants.LOAD_EXTERNAL_DTD_FEATURE);
```

Comment:
Shouldn't we throw and stop the flow here? Say the first line in the try block threw an error, and since we are continuing the flow, all other properties will not be set AFAIU.

Reply:
I don't think we need to stop the flow here. Because what we are doing here is setting some properties to securely process the xml file. If setting these properties failed, that doesn't mean we can't read the file. So we can continue the process. Only thing is some of the security properties will not be set. Which is fine IMO.